precise/iaf_psc_exp_ps_lossless.cpp

```diff
@@ -0,0 +1,671 @@
+  }
+  else
+  {
+    U_reset_ -= delta_EL;
+  }
+
+  if ( U_reset_ >= U_th_ )
+  {
+    throw BadProperty( "Reset potential must be smaller than threshold." );
+  }
+
+  if ( U_reset_ < U_min_ )
+  {
+    throw BadProperty(
+      "Reset potential must be greater equal minimum potential." );
+  }
+
+  if ( c_m_ <= 0 )
+  {
+    throw BadProperty( "Capacitance must be strictly positive." );
+  }
+
+  if ( t_ref_ < 0 )
+  {
+    throw BadProperty( "Refractory time must not be negative." );
+  }
+
+  if ( tau_m_ <= 0 || tau_ex_ <= 0 || tau_in_ <= 0 )
+  {
+    throw BadProperty( "All time constants must be strictly positive." );
+  }
+
+  if ( tau_m_ == tau_ex_ || tau_m_ == tau_in_ )
+  {
+    throw BadProperty(
+      "Membrane and synapse time constant(s) must differ."
+      "See note in documentation." );
+  }
+
+  return delta_EL;
+}
+
+void
+nest::iaf_psc_exp_ps_lossless::State_::get( DictionaryDatum& d,
+  const Parameters_& p ) const
+{
+  def< double >( d, names::V_m, y2_ + p.E_L_ ); // Membrane potential
+  def< bool >( d, names::is_refractory, is_refractory_ );
+  def< double >(
+    d, names::t_spike, Time( Time::step( last_spike_step_ ) ).get_ms() );
+  def< double >( d, names::offset, last_spike_offset_ );
+  def< double >( d, names::I_syn_ex, I_syn_ex_ );
+  def< double >( d, names::I_syn_in, I_syn_in_ ); // y1 state
+  def< double >( d, names::I_syn, I_syn_ex_ + I_syn_in_ );
+}
+
+void
+nest::iaf_psc_exp_ps_lossless::State_::set( const DictionaryDatum& d,
+  const Parameters_& p,
+  double delta_EL )
+{
+  if ( updateValue< double >( d, names::V_m, y2_ ) )
+  {
+    y2_ -= p.E_L_;
+  }
+  else
+  {
+    y2_ -= delta_EL;
+  }
+
+  updateValue< double >( d, names::I_syn_ex, I_syn_ex_ );
+  updateValue< double >( d, names::I_syn_in, I_syn_in_ );
+}
+
+/* ----------------------------------------------------------------
+ * Default and copy constructor for node
+ * ---------------------------------------------------------------- */
+
+nest::iaf_psc_exp_ps_lossless::iaf_psc_exp_ps_lossless()
+  : Archiving_Node()
+  , P_()
+  , S_()
+  , B_( *this )
+{
+  recordablesMap_.create();
+}
+
+nest::iaf_psc_exp_ps_lossless::iaf_psc_exp_ps_lossless(
+  const iaf_psc_exp_ps_lossless& n )
+  : Archiving_Node( n )
+  , P_( n.P_ )
+  , S_( n.S_ )
+  , B_( n.B_, *this )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Node initialization functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::iaf_psc_exp_ps_lossless::init_state_( const Node& proto )
+{
+  const iaf_psc_exp_ps_lossless& pr =
+    downcast< iaf_psc_exp_ps_lossless >( proto );
+
+  S_ = pr.S_;
+}
+
+void
+nest::iaf_psc_exp_ps_lossless::init_buffers_()
+{
+  B_.events_.resize();
+  B_.events_.clear();
+  B_.currents_.clear(); // includes resize
+  B_.logger_.reset();
+}
+
+void
+nest::iaf_psc_exp_ps_lossless::calibrate()
+{
+  B_.logger_
+    .init(); // ensures initialization in case mm connected after Simulate
```

Comment:
This would be easier to read for me if the comment is placed above the `B_.logger_.init();` so that the actual code is one line.

Reply:
Absolutely.